02_oop/Src/mainwindow.cpp

```diff
@@ -0,0 +1,271 @@
+#include "mainwindow.h"
+#include "./ui_mainwindow.h"
+
+MainWindow::MainWindow(QWidget* parent)
+    : QMainWindow(parent)
+    , ui(new Ui::MainWindow)
+{
+    ui->setupUi(this);
+    centralWidget = new QWidget(this);
+    this->setCentralWidget(centralWidget);
+    map = new QGridLayout(centralWidget);
+
+    Reset = new QPushButton();
+    Reset->setText("New / Reset");
+    QObject::connect(Reset, SIGNAL(clicked()), this, SLOT(ResetGame()));
+    map->addWidget(Reset, 0, 0);
+
+    scoreText = 0;
+    scoreText = new QLabel("Score: \n 0", this);
+    scoreText->setAlignment(Qt::AlignCenter);
+    scoreText->setStyleSheet("QLabel { background-color: #bbae9e; color: white; font:15pt; "
+                             "font-weight:200; border-radius: 5px; }");
+    map->addWidget(scoreText, 0, 3);
+
+    for (int i = 0; i < 4; ++i)
+        for (int j = 0; j < 4; ++j)
+        {
+            cell[i][j] = new QLabel("", this);
+            map->addWidget(cell[i][j], i + 1, j);
+            cell[i][j]->setAlignment(Qt::AlignCenter);
+            Color(i, j);
+        }
+    cell[1][0]->setFocus();
+    setWindowTitle("The Game 2048");
+    setFixedSize(500, 500);
+    Generate();
+}
+
+MainWindow::~MainWindow()
+{
+    delete ui;
+}
+
+void MainWindow::ResetGame()
+{
+    for (int i = 0; i < 4; ++i)
+        for (int j = 0; j < 4; ++j)
+        {
+            cell[i][j]->setText("");
+            Color(i, j);
+        }
+    score = 0;
+    scoreText = 0;
+    scoreText = new QLabel("Score: \n 0", this);
+    scoreText->setAlignment(Qt::AlignCenter);
+    scoreText->setStyleSheet("QLabel { background-color: #bbae9e; color: white; font:15pt; "
+                             "font-weight:200; border-radius: 5px; }");
+    map->addWidget(scoreText, 0, 3);
+    Reset->clearFocus();
+    Generate();
+}
+
+void MainWindow::Generate()
+{
+    bool hasPlace = false;
+    for (int i = 0; i < 4; ++i)
+        for (int j = 0; j < 4; ++j)
+        {
+            if (cell[i][j]->text() == "" || cell[i][j]->text() == "2")
+            {
+                hasPlace = true;
+                break;
+            }
+        }
+    if (!hasPlace)
+    {
+        QDialog* endDialog = new QDialog();
+        endDialog->setMinimumWidth(200);
+        endDialog->show();
+    }
+    int row = 0, column = 0;
+    do
+    {
+        row = rand() % 4;
+        column = rand() % 4;
+    } while (cell[row][column]->text() != "" && cell[row][column]->text() != "2");
+    cell[row][column]->setText(QString::number(cell[row][column]->text().toInt() + 2));
+    Color(row, column);
+}
+
+void MainWindow::Color(int i, int j)
+{
+    switch (cell[i][j]->text().toInt())
```

Comment:
Can this switch be simplified by calculating of style based on power of 2?
Linear complexity isn't best solution. Especially, taking into account that the original game supports numbers greater than 2048.

Reply:
I didn't set a goal for a count greater than 2048.
If more, what algorithm should be used to change the color of the cells?